### ISSUE TITLE
Integrate Wikipedia Data

### ISSUE THREAD

**Issue by TheOnlyWayUp** [sarkartanmay393/GeoPulse#6]
Hey Tanmay, cool project!

Atm the data is from GPT-4o, https://github.com/sarkartanmay393/GeoPulse/blob/11d926a97eace592b28d8489f9014a6ab0536cfd/src/app/api/generate/route.ts#L46-L78, that's cool - but it's not dynamic.

Maybe you could integrate data from the Wikipedia pages (which change as events occur, geopolitical and otherwise), and feed it into GPT for better information (something like RAG?).

Example: [France-India Relations | wikipedia.org](https://en.wikipedia.org/wiki/France–India_relations)

**Comment by sarkartanmay393** [comment]
@TheOnlyWayUp Yes definitely, I can do that. Your idea is on my list now. If you are free sometimes, we can chat about it.

**Comment by sarkartanmay393** [comment]
This can be used, Public API: https://api.wikimedia.org/wiki/Core_REST_API/Reference/Search/Search_result_object

**Comment by Nil2000** [comment]
For Future ref: [China_India_relation](https://en.wikipedia.org/w/rest.php/v1/page/China%E2%80%93India_relations/html)
HTML format

**Comment by sarkartanmay393** [comment]
@Nil2000 Do test the initial wikipedia attachment. Now we are passing the whole extracted text from the html to chatgpt.  Passing max 15k tokens.

**Comment by TheOnlyWayUp** [comment]
> @Nil2000 Do test the initial wikipedia attachment. Now we are passing the whole extracted text from the html to chatgpt. Passing max 15k tokens.

Hmm, I took a look at Wikipedia's simple pages (simple.wikipedia.org), they have very few stubs for geopolitical relations (valid, they're hard to boil down in most cases).

I see a few options, ranked by how well they'd do (imo)
1. Summarizing each heading and its content, bringing down the token usage (but requires more requests).
2. Embedding each heading (and its content) and comparing them to each of your 6 parameters. The top `k` results would be provided to the model instead of the entire article. (This is an RAG Approach)
3. Truncating everything past the first paragraph for each heading. (Worth a try, LLMs can perform well with limited information.)

Tias!

**Comment by ghost** [comment]
> > @Nil2000 Do test the initial wikipedia attachment. Now we are passing the whole extracted text from the html to chatgpt. Passing max 15k tokens.
> 
> Hmm, I took a look at Wikipedia's simple pages (simple.wikipedia.org), they have very few stubs for geopolitical relations (valid, they're hard to boil down in most cases).
> 
> I see a few options, ranked by how well they'd do (imo)
> 1. Summarizing each heading and its content, bringing down the token.
> 2. Embedding each heading (and its content) and comparing them to each of your 6 parameters. The top `k` results would be provided to the model instead of the entire article. (This is an RAG Approach)
> 3. Truncating everything past the first paragraph for each heading. (Worth a try, LLMs can perform well with limited information.)
> 
> Tias!

You are right, using the RAG approach is the way to go for cost effective and more accuracy. 

And also for more current results if we fetch top 20-30 news articles from the internet of selected two countries and use that if it would be much much better but it would cost more. 

Ideal: Use Wikipedia + GPT for overall knowledge, and latest news.

**Comment by sarkartanmay393** [comment]
Added a quick support for this, but need to improve on it further.

**Comment by tanmay-retwin** [comment]
https://github.com/sarkartanmay393/GeoPulse/issues/6
